haystack/tools/component_tool.py

```diff
@@ -0,0 +1,296 @@
+
+from haystack import logging
+from haystack.core.component import Component
+from haystack.tools import Tool
+
+logger = logging.getLogger(__name__)
+
+
+def is_pydantic_v2_model(instance: Any) -> bool:
+    """
+    Checks if the instance is a Pydantic v2 model.
+
+    :param instance: The instance to check.
+    :returns: True if the instance is a Pydantic v2 model, False otherwise.
+    """
+    return hasattr(instance, "model_validate")
+
+
+class ComponentTool(Tool):
+    """
+    A Tool that wraps Haystack components, allowing them to be used as tools by LLMs.
+
+    ComponentTool automatically generates LLM-compatible tool schemas from Component input sockets,
+    which are derived from the component's `run` method signature and type hints.
+
+
+    Key features:
+    - Automatic LLM tool calling schema generation from Component input sockets
+    - Type conversion and validation for Component inputs
+    - Support for complex types (dataclasses, Pydantic models, lists)
+    - Automatic name generation from Component class name
+    - Description extraction from Component docstrings
+
+    Let's assume we already have (or want to create) a Haystack component that we want to use as a tool.
+    We can create a ComponentTool from the component by passing the component to the ComponentTool constructor.
+
+    ```python
+    from haystack import component
+    from haystack.tools import ComponentTool
+
+    @component
+    class WeatherComponent:
+        '''Gets weather information for a location.'''
+
+        def run(self, city: str, units: str = "celsius"):
+            '''
+            :param city: The city to get weather for
+            :param units: Temperature units (celsius/fahrenheit)
+            '''
+            return f"Weather in {city}: 20°{units}"
+
+    # Create a tool from the component
+    weather = WeatherComponent()
+    tool = ComponentTool(
+        component=weather,
+        name="get_weather",  # Optional: defaults to snake_case of class name
+        description="Get current weather for a city"  # Optional: defaults to component run method docstring
+    )
+    ```
+
+    """
+
+    def __init__(self, component: Component, name: Optional[str] = None, description: Optional[str] = None):
+        """
+        Create a Tool instance from a Haystack component.
+
+        :param component: The Haystack Component to wrap as a tool
+        :param name: Optional name for the tool (defaults to snake_case of Component class name)
+        :param description: Optional description (defaults to Component's docstring)
+        :raises ValueError: If the component is invalid or schema generation fails
+        """
+        if not isinstance(component, Component):
+            message = (
+                f"Object {component!r} is not a Haystack component. "
+                "Use this method to create a Tool only with Haystack component instances."
+            )
+            raise ValueError(message)
+
```

Comment:
Here I am contradicting my comment in the experimental PR (https://github.com/deepset-ai/haystack-experimental/pull/159/files/ce864dd8f4f10c2c42f3ded35a515ef71a995585#r1892494701)
but probably now that we have clarified that components to convert into tools MUST NOT be part of a Pipeline, I think it is worth adding a check.

Reply:
thanks!

Reply:
Let's add a simple test to cover this.